airbyte-cdk/python/airbyte_cdk/connector_builder/migrator/source.py

```diff
@@ -0,0 +1,166 @@
+                manifest = yaml.safe_load(f.read())
+            if "spec" in manifest:
+                # It should be fine to ignore since the CATs test `test_match_expected` ensure that both are the same
+                logger.info(
+                    f"Source {source_name} has both spec file and spec within manifest. Will keep manifest as-is and delete the spec file"
+                )
+            else:
+                manifest["spec"] = spec.dict(exclude_none=True)
+
+            with open(manifest_path, 'w') as outfile:
+                yaml.dump(manifest, outfile, sort_keys=False, Dumper=_ManifestDumper)
+            self._delete_spec_outside_manifest(source_name)
+        else:
+            logger.debug(f"Source {source_name} does not have a spec outside the manifest.yaml")
+
+    def _fetch_spec_outside_manifest(self, source_name: str) -> Optional[Spec]:
+        json_spec_path = os.path.join(self._python_package_path(source_name), "spec.json")
+        yaml_spec_path = os.path.join(self._python_package_path(source_name), "spec.yaml")
+
+        if os.path.exists(json_spec_path):
+            with open(json_spec_path) as json_file:
+                return self._assemble_spec(json.load(json_file))
+        elif os.path.exists(yaml_spec_path):
+            with open(yaml_spec_path) as yaml_file:
+                return self._assemble_spec(yaml.safe_load(yaml_file))
+        else:
+            return None
+
+    @staticmethod
+    def _assemble_spec(spec: Dict[str, Any]) -> Spec:
+        return Spec(
+            type="Spec",
+            connection_specification=spec["connectionSpecification"],
+            documentation_url=spec.get("documentationUrl", None),
+            advanced_auth=AuthFlow(**spec["advanced_auth"]) if "advanced_auth" in spec else None
+        )
+
+    def _delete_spec_outside_manifest(self, source_name: str) -> None:
+        json_spec_path = os.path.join(self._python_package_path(source_name), "spec.json")
+        yaml_spec_path = os.path.join(self._python_package_path(source_name), "spec.yaml")
+
+        if os.path.exists(json_spec_path):
+            os.remove(json_spec_path)
+        elif os.path.exists(yaml_spec_path):
+            os.remove(yaml_spec_path)
+
+    def fetch_no_code_sources(self) -> List[str]:
+        no_code_sources = []
+
+        connectors_folder_path = os.path.join("airbyte-integrations", "connectors")
+        for directory in os.listdir(connectors_folder_path):
+            manifest_file_path = os.path.join(connectors_folder_path, directory, directory.replace("-", "_"), "manifest.yaml")
+            if self._is_no_code_connector(manifest_file_path):
+                no_code_sources.append(directory)
+        return no_code_sources
+
+    @staticmethod
+    def _is_no_code_connector(manifest_file_path: str) -> bool:
+        if not os.path.exists(manifest_file_path):
+            return False
+
+        with open(manifest_file_path) as manifest_file:
```

Comment:
The code needs to be improved as some sources have the frame of no_code but do not implement `YamlDeclarativeSource` (example: sendgrid)

Reply:
do you know of others where that may be the case? For context, Sendgrid is a particularly special case in that we had migrated it as low-code but we reverted it back because we weren't super confident that all the edge cases had been fixed. We left the code in as a reference point for future migrations. I'd consider this more of an in-development connector since we're not even sure it works anymore. And I don't think it is a big priority to solve for it if this is the only one

Reply:
I don't, no. This is the one that raised an issue on the migration. If for spec, it might be fine and we can put the spec in the manifest.yaml but this will not work for the schemas as we will delete the "schemas" folder which source-sendgrid relies on